File: Java/jakarta/filesystemstorage/src/main/java/com/ithit/webdav/samples/fsstorageservlet/extendedattributes/ExtendedAttributeFactory.java
package com.ithit.webdav.samples.fsstorageservlet.extendedattributes;

/**
 * Factory-singleton which creates a ExtendedAttribute instance.
 * Instance is valid for the current platform.
 */
final class ExtendedAttributeFactory {

    private ExtendedAttributeFactory() {
    }

    private static ExtendedAttribute extendedAttribute;

    /**
     * Builds a specific ExtendedAttribute for the current platform.
     *
     * @return Platform specific instance of ExtendedAttribute.
     */
    static synchronized ExtendedAttribute buildFileExtendedAttributeSupport() {
        if (extendedAttribute == null) {
            if (System.getProperty("os.name").toLowerCase().contains("mac")) {
                extendedAttribute = new OSXExtendedAttribute();
            } else {
                extendedAttribute = new DefaultExtendedAttribute();
            }
        }
        return extendedAttribute;
    }

}

File: Kotlin/kotlinfs/src/main/kotlin/com/ithit/webdav/samples/kotlinfs/extendedattributes/ExtendedAttributeFactory.kt
package com.ithit.webdav.samples.kotlinfs.extendedattributes

/**
 * Factory-singleton which creates a ExtendedAttribute instance.
 * Instance is valid for the current platform.
 */
internal object ExtendedAttributeFactory {

    private var extendedAttribute: ExtendedAttribute? = null

    /**
     * Builds a specific ExtendedAttribute for the current platform.
     *
     * @return Platform specific instance of ExtendedAttribute.
     */
    @Synchronized
    fun buildFileExtendedAttributeSupport(): ExtendedAttribute? {
        if (extendedAttribute == null) {
            extendedAttribute = if (System.getProperty("os.name").toLowerCase().contains("mac")) {
                OSXExtendedAttribute()
            } else {
                DefaultExtendedAttribute()
            }
        }
        return extendedAttribute
    }

}
